### ISSUE TITLE
Error when using restart files

### ISSUE THREAD

**Issue by arbennett** [NCAR/summa#253]
When trying to run SUMMA using the initial conditions as a restart file from a previous run I get an error:
```
[andrbenn@n0561 06305-06320]$ /civil/hydro/arbennett/summa/bin/summa_debug -m summa_fileManager.txt -g 6305 16 -p m -r y                                                              
file_master is 'summa_fileManager.txt'.
 GRU-Parallelization run activated. 16 GRUs are selected for simulation.
start at 14:09:13
Name of Model Output control file: /civil/hydro/michaelou/summaProj/summaData/blivneh/summa_run/oc_columbia_v2.txt
forrtl: severe (408): fort: (2): Subscript #1 of the array SNOWDATA has value 6305 which is greater than the upper bound of 16

Image              PC                Routine            Line        Source             
summa_debug        0000000000A5B166  Unknown               Unknown  Unknown
summa_debug        000000000089FA33  read_icond_module          95  read_icond.f90

Stack trace terminated abnormally.
```

I believe (but have not yet confirmed) that this is due to using the GRU Parallelization mode (ie with the `-g 06305 16` option).

**Comment by arbennett** [comment]
The issue appears to be that the `hruID` is not being written out correctly.  With the current implementation the "correct" way to use a restart would be to specify `-g 1 16` since the written HRU index starts at 1.  This is not the ideal behavior.

**Comment by bartnijssen** [comment]
Problems:
* The restart file does not contain a `gru` dimension or `gruId`
* The restart file does contain a `hru` dimension but not a `hruId`
* The `-g --gru           Run a subset of countGRU GRUs starting from index startGRU` command-line option to SUMMA does not properly parse the GRU. There are two problems:
  + The specified values are not treated as IDs, but simply used as array indices in the code. We should actually provide a lookup that maps the IDs specified on the command-line to the correct elements in the model setup
  + While the option specifies that these are GRUs, the command-line value is used as an index into an array of HRUs. The implicit assumption here is that GRU and HRU IDs are aligned and that there is only one HRU per GRU.

Solutions
* Add gru dimension and gruId to restart file
* Add hruId to restart file
* Parse gruId and hruId when reading the restart file
* Create a lookup function that maps an hruId and gruId to the correct model elements
* Use this same lookup to map the gruId specified on the command-line to the correct model elements

@arbennett: Please comment whether this makes sense

**Comment by arbennett** [comment]
I'm not sure on the third point in your problems section, but the first two are definitely problems. It's possible that when using `-g startGRU nGRU` it does look for the right numbered HRU (which would still be wrong) but the restart files are not indexed correctly.

**Comment by bartnijssen** [comment]
Yeah - the third one is definitely a problem as well. We should keep GRUs and HRUs cleanly separated and not use the former as an index for the latter. It'll be hopelessly confusing (as demonstrated by the discussion).

**Comment by bartnijssen** [comment]
Note - when performing restart tests for model setups with single HRUs:
* restarts are exact (bit-for-bit) in the absence of snow
* restarts are close (but not exact) when snow is present in the state file

**Comment by andywood** [comment]
Good to get these issues addressed.  I've been using restart files successfully, but working around the lack of functionality from not having comprehensive ID info in the state files (ie one can either run whole domain or single GRU runs).  Fortunately there's a lot of existing code in SUMMA to borrow to do this (eg mapping/lookup).

**Comment by andywood** [comment]
as of PR #430 'gru' and also 'tdh' are new dimensions in the state file.  these allow for writing basin hillslope runoff variables.

**Comment by ashleymedin** [comment]
We have  hruId and gruId  written in the restarts with a call to write_hru_info, https://github.com/CH-Earth/summa/blob/develop_sundials/build/source/netcdf/modelwrite.f90, and these are checked if available in https://github.com/CH-Earth/summa/blob/develop_sundials/build/source/netcdf/read_icond.f90 or assumed the same as the attributes file if not there for backwards compatibility. We do not yet have the -g and -h flag actually refer to the gruId and not the index in the attribute file as suggested, probably should change that.

**Comment by ashleymedin** [comment]
Goal, as I see it, from the point we are at now with hruId and gruId written to the restarts if you have a new code restart @andywood @wknoben @martynpclark:
If I have input files containing all of N.Amer and only run 1 GRU, and make a restart, I want to be able to run the 1 GRU from that restart even though now I don’t have all the GRUs in the restart but I do in the attributes and trial params. Right now, if I have all of N. Amer and want to run 1 GRU (or a subset) and restart it, I have to subset the attributes and trial params to that 1 GRU. Similarly for a subset of GRUs and a restart. 

The hardest part is trying to get it so it can also be backwards compatible. In order to make it backwards compatible, for example in the example by @arbennett above, if we first ran with `-g 06305 16` and made a restart, we could run the restart with `-g 06305 16` if I fix the code, but IF:

1. the user does `-g 06305 16` to run the restart and `gruId` is not in the restart file we know it is the old version of the code the code that made this restart so this should be an error.
2. if the user does `-g 1 16` and `gruId` is not in the restart file it should assume `1` refers to the gru index. If the attributes and trial params weren't also subset (so the same number of GRUs are in the restart as the attributes, this should be an error since it won't know the index in the attributes and trial params.
3. if `gruId` is in the restart file but there's no `gruId=1` it should assume `1` refers to the the gru index in the restart and match the gru index to the gruId to get the attributes. 

I could similarly update the `-h` mode, which maybe has never worked? https://github.com/CH-Earth/summa/issues/351
